azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +73,176 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+            return self.__parse_policy(custom_policy)
+
+    @staticmethod
+    def __parse_policy(custom_policy):
+        """
+        Return a policy combining the default and custom_policy. Any attributes provided in the custom policy
+        override default values used in the template. If attribute is not provided, default is used.
+
+        The value of the "extensions" attribute is a case-folded dict. CRP allows extensions to be any case, so we use
+        case-folded dict to allow for case-insensitive lookup of individual extension policies.
+
+        The expected custom policy format is:
+         {
+            "policyVersion": str,
+            "extensionPolicies": {
+                "allowListedExtensionsOnly": bool,
+                "signatureRequired": bool,
+                "extensions": {
+                    "<extensionName1>": {
+                        "signatureRequired": bool
+                    }
+                }
+            }
+        }
+        """
+        def __validate_attribute_type(attribute_name, attribute_value, expected_type, parent_attribute_name=None):
+            """
+            Helper method to raise PolicyInvalidError with useful error message.
+            """
+            if not isinstance(attribute_value, expected_type):
+                # Error message should refer to "JSON" not "dict".
+                if expected_type == dict:
+                    expected_type_in_msg = "JSON"
+                else:
+                    expected_type_in_msg = expected_type.__name__
+
+                if parent_attribute_name is None:
+                    msg = ("invalid type {0} for attribute '{1}', please change to {2}."
+                           .format(type(attribute_value).__name__, attribute_name, expected_type_in_msg))
+                else:
+                    msg = ("invalid type {0} for attribute '{1}' in section '{2}', please change to {3}."
+                           .format(type(attribute_value).__name__, attribute_name, parent_attribute_name, expected_type_in_msg))
+
+                raise PolicyInvalidError(msg)
+
+        # We use the default policy as a template, and replace any attributes provided in the custom policy.
+        template = \
+            {
+                "policyVersion": "0.1.0",
+                "extensionPolicies": {
+                    "allowListedExtensionsOnly": _DEFAULT_ALLOW_LISTED_EXTENSIONS_ONLY,
+                    "signatureRequired": _DEFAULT_SIGNATURE_REQUIRED,
+                    "extensions": {}
+                }
+            }
+
+        extension_policies = custom_policy.get("extensionPolicies")
+        if extension_policies is not None:
+
+            # Validate and update template with global allowlist policy
+            allowlist_policy = extension_policies.get("allowListedExtensionsOnly")
+            if allowlist_policy is not None:
+                __validate_attribute_type("allowListedExtensionsOnly", allowlist_policy, bool)
```

Comment:
signatureRequired is present both in the "extensionPolicies" attribute (global section) and the "extensions" attribute (individual extension section). We specify the parent attribute just to avoid this ambiguity in the error message. allowListedExtensionsOnly is only present in "extensionPolicies" so we don't need to specify the parent.